### PR TITLE
Fix two regressions since 0.1.0

### PR DESCRIPTION
**Pull request by pepeiborra**
I noticed these running VSCode + ghcide master on the GHC codebase

### CODE REVIEW

**Comment by mpickering** [comment]
How did you observe these?

**Comment by pepeiborra** [comment]
> How did you observe these?

Shake profiling showed an unexpected large number of `GetFileExists` calls for hovers. I ran into the second one anecdotally during workspace reloads

**Comment by ndmitchell** [comment]
FWIW, I noticed the issue with hie-bios earlier today, and raised the upstream issue at https://github.com/snoyberg/yaml/issues/185 - since that's the most common way I had the hie-bios initialisation failing. Agreed this is a good fix though! I do wonder if uninteruptibleMask is required though, since everywhere I've used mask alone, sooner or later I've regretted it.

**Comment by mpickering** [comment]
I have also been seeing that issue, thanks for tracking it down Neil and Pepe.

**Comment by pepeiborra** [comment]
> FWIW, I noticed the issue with hie-bios earlier today, and raised the upstream issue at [snoyberg/yaml#185](https://github.com/snoyberg/yaml/issues/185) - since that's the most common way I had the hie-bios initialisation failing. Agreed this is a good fix though! I do wonder if uninteruptibleMask is required though, since everywhere I've used mask alone, sooner or later I've regretted it.

I don't think this is the same issue though. The problem I'm trying to fix is the `LoadCradle` retry happening before the OS has had a chance to reclaim the resources used by the external processes started by hie-bios. In the case of Hadrian, the Shake file lock. Before the fix, this was failing as below.

About `mask` vs `uninterruptibleMask`, I'm happy to follow your advice here.

```
Started LSP server in 0.30s
[DEBUG] Set files of interest to: [NormalizedFilePath "/home/pepe/scratch/ghc/compiler/main/GHC.hs"]
[DEBUG] Starting shakeRun (aborting the previous one took 0.00s)
[INFO] Opened text document: file:///home/pepe/scratch/ghc/compiler/main/GHC.hs
[DEBUG] Running cradle /home/pepe/scratch/ghc/hie.yaml
[DEBUG] Starting shakeRun (aborting the previous one took 0.00s)
[DEBUG] Finishing shakeRun (took 0.04s, exception: AsyncCancelled)
[DEBUG] Running cradle /home/pepe/scratch/ghc/hie.yaml
> hadrian: Shake failed to acquire a file lock on .hie-bios/hadrian/.shake.lock
> Shake process ID 31573 is using this lock.
> setLock: resource exhausted (Resource temporarily unavailable)
> CallStack (from HasCallStack):
>   errorIO, called at src/General/FileLock.hs:67:9 in shake-0.18.3-af1d490ec7e20705e035a339d81952289a7a6aad11ff39a6e3d7c78f9a6beb32:General.FileLock
```

File: test/exe/Main.hs [hunk 1]
```diff
@@ -430,13 +430,13 @@ codeLensesTests = testGroup "code lenses"
...
   [ testSession "workspace file" $ do
-      _ <- openDoc' "A.hs" "haskell" "module A where"
+      _ <- openDoc' "C/A.hs" "haskell" "module A where"
       RequestMessage{_params = RegistrationParams (List regs)} <- skipManyTill anyMessage (message @RegisterCapabilityRequest)
       let watchedFileRegs =
             [ args | Registration _id WorkspaceDidChangeWatchedFiles args <- regs ]
       liftIO $ assertBool "watches workspace files" $ not $ null watchedFileRegs
   , testSession "non workspace file" $ do
-      _ <- openDoc' "/tmp/A.hs" "haskell" "module A where"
+      _ <- openDoc' "/tmp/A.hs" "haskell" "{-# LANGUAGE NoImplicitPrelude#-}\nmodule A where"
```
Comment (cocreature) [inline test/exe/Main.hs]: What changed to require this?

File: test/exe/Main.hs [hunk 2]
```diff
@@ -430,13 +430,13 @@ codeLensesTests = testGroup "code lenses"
 watchedFilesTests :: TestTree
 watchedFilesTests = testGroup "watched files"
   [ testSession "workspace file" $ do
-      _ <- openDoc' "A.hs" "haskell" "module A where"
+      _ <- openDoc' "C/A.hs" "haskell" "module A where"
```
Comment (cocreature) [inline test/exe/Main.hs]: Why did this change?

File: test/exe/Main.hs (hunk 1 above)
Comment (pepeiborra) [reply test/exe/Main.hs]: Now there is a `DidChangeWatchedFile` subscription message for `Prelude` on various paths, which were not emitted before as they would use relative paths.

Comment (cocreature) [reply test/exe/Main.hs]: I’m not following, why do we get a `DidChangeWatchedFile` notification for `Prelude`? Surely we are not editing the contents of that file in this test?

File: test/exe/Main.hs (hunk 2 above)
Comment (pepeiborra) [reply test/exe/Main.hs]: This is an incomplete attempt at making a test more demanding. Will try to push an improved version tonight.

File: test/exe/Main.hs (hunk 1 above)
Comment (pepeiborra) [reply test/exe/Main.hs]: `DidChangeWatchedFile` subscriptions are used to implement the `GetFileExists` Shake rule, which is used by `GetLocatedImports` and other Shake rules. So it's not just a matter of whether we are editing the contents of that file in the test.

File: test/exe/Main.hs (hunk 2 above)
Comment (pepeiborra) [reply test/exe/Main.hs]: pushed

File: test/exe/Main.hs (hunk 1 above)
Comment (cocreature) [reply test/exe/Main.hs]: Oh sorry, I mixed up `DidChangeTextDoucment` and `DidChangeWatchedFiles`. Makes sense now :+1:

**Comment by cocreature** [comment]
looks like this needs rebasing on master

**Comment by pepeiborra** [comment]
Rebased and force pushed. By the way, I ended up not following the suggestion to use `uninterruptibleMask` as `mask` seems enough and the Haddocks for `uninterruptibleMask` are very scary - I don't want to make ghcide unkillable if e.g. the wrong craddle is selected and stack is building the world.

File: test/exe/Main.hs (hunk 1 above)
Comment (alanz) [reply test/exe/Main.hs]: I am hitting this problem in haskell-language-server tests too.  Is the correct solution not to filter out the watch request on non-existent Prelude files?

Comment (cocreature) [reply test/exe/Main.hs]: @alanz no, we need to detect if a `Prelude` file starts existing.

Comment (alanz) [reply test/exe/Main.hs]: @cocreature The file monitored as the location of the implicit prelude file is assumed to be in the same directory as the original file. Or in some unlikely location, like deep in the stack workspace, with the project subdirectory appended.

So does it make sense to monitor for that file? Would it not make more sense to monitor wherever the implicit prelude actually exists/comes from?  If that file does suddenly spring into existence, will it replace the implicit prelude?

Comment (cocreature) [reply test/exe/Main.hs]: Prelude isn’t all that special in GHC. E.g., you can use `base-noprelude` and provide your own `Prelude` module and GHC will pick that. So I don’t think it makes sense to special case prelude in any way.

**Pull request merged**